Implement a basic Subcommand structure.

Aathish04 opened the pull request:
<!--- 
Thanks for contributing to manim!
**Please ensure that your pull request works with the latest version of manim from this repository.**
You should also include:
  1. The motivation for making this change (or link the relevant issues)
  2. How you tested the new behavior (e.g. a minimal working example, before/after
     screenshots, gifs, commands, etc.) This is rather informal at the moment, but
     the goal is to show us how you know the pull request works as intended.
If you don't need any of the optional sections, feel free to delete them to prevent clutter.
-->

## List of Changes
- Add a subparser for commands related to modifying the config files manim uses and creates.
- Implement `write`, `show` and `export` commands with said subparser. 
  - **Write**
      Writes a custom config file based off of `default.cfg` and saves it to the user config location if the level passed is `user` or to the current working directory if the level passed is `cwd`. If no level is passed, the config file writer will ask the user where to save the config.
      ```bash
      manim cfg write --level cwd
      manim cfg write --level user
      manim cfg write
      ```
  - **Show**
      Prints the Final Configuration at the current working directory to the Terminal.
      ```bash
      manim cfg show
      ```
  - **Export**
      Exports the Final Configuration at the current working directory to --dir.
      ```bash
      manim cfg export
      manim cfg export --dir ./CustomConfigs/
      ```
- Remove the `manim-cfg` command.

## Motivation
The introduction of the Config File writer opened up a lot of possibilities for config related utilities as discussed in #201 . These utilities would go well under the `manim` command, under the `cfg` subcommand, as suggested by @leotrs, and that is what this PR addresses.

## Explanation for Changes
A subparser for commands related to modifying the config files manim uses and creates has been added, and to this, 3 more sub-subparsers, for the `write`, `show` and `export` subcommands have been added.

The `manim-cfg` command has been removed, as it only linked to a single function from a single file, and implementing subcommands for every variant of `manim-cfg` and `manimcm-cfg` would get very tedious. This command has been wholly replaced by the subcommands to manim mentioned previously.


## Testing Status
Tested with no errors, and outputs as expected.

![Page1](https://user-images.githubusercontent.com/33193764/88206585-22c39a00-cc6c-11ea-88be-c533ed32cdc6.png)
![Page2](https://user-images.githubusercontent.com/33193764/88206610-29eaa800-cc6c-11ea-9377-bf456bc12fd4.png)


## Further Comments
I'm not entirely satisfied with the way I've implemented this. Currently, the `__main__` function in `main.py` has to be modified to not ask for a default file if the `cfg` subcommand is invoked, and I feel like my solution was a tad too hacky. If a better way exists, don't hesitate to review this PR to shreds.

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

<!-- Once again, thanks for helping out by contributing to manim! -->

Aathish04 commented:
For the benefit of those who want to review this, I'd like to give a rundown of how the subcommand system will work. 

- _In the beginning_, calling `manim` from the command line will run `manim/main.py:__main__`.
In `main.py` `utils/cfg_file_utils` is imported, which has the functions that all the subcommands will execute.
Before the function is executed, `file_writer_config` and `args` are imported from `config.py`.

    - In `config.py`,` _run_config()` is imported from `config_utils.py`.
    - In `_run_config()`, `_parse_cli()` is called. 
        - Here, an explicit check is performed to see if a subcommand has been executed along with the main `manim` command. 
        - If it _does_ find a subcommand, the normal parser gets a shiny new subparser, within which a sub-subparser for cfg related commands is intialised and arguments for those added. Note that if it finds a subcommand, `file` and other compulsory arguments won't be added to the main Argument Parser. If it _doesn't_ find a subcommand, it realises that this is just to render an animation, and everything is initialised "normally". 
        - Now, the cfg sub-subparser parses the subcommand for cfg to determine what to do, and _manually_ adds it to the main parser's Namespace. `_run_config()` then returns the main parser's Namespace.
    - Now that `_run_config()` has the required namespace (called `args` here, for brevity), it will check if `args` has a `subcommands` attribute. 
If it does, it gets the cwd's `manim.cfg` if it's present, and adds it to the list of config_files. If it doesn't, the usual steps that were taken before this PR are followed.
    - `_run_config` then calls `_parse_file_writer_config` to get `file_writer_config`.
        - If `args` has a `subcommands` attribute, `_parse_file_writer_config` won't set `input_file` and other parameters that _don't_ matter in a subcommand in fw_config.
        - If `args` doesn't have such an attribute, `_parse_file_writer_config` proceeds as normal and returns the `file_writer_config`.

    - All of these configs, along with a list of the parsed cfg files returned by `_run_config` is unpacked in `config.py`.

- Back in `__main__.py`, if `args` has any subcommands, those subcommands are given top priority and the functions corresponding to them are called.  If they don't, then `manim` assumes it's just being asked to render an animation and does that.

And that's how this subcommand structure works.

naveen521kk commented on manim/config.py:
```diff
@@ -89,7 +89,7 @@ def _parse_config(config_parser, args):
 args, config_parser, file_writer_config, successfully_read_files = _run_config()
 if _from_command_line():
-    logger.info(f"Read configuration files: {os.path.abspath(successfully_read_files[-1])}")
+    logger.info(f"Read configuration files: {[os.path.abspath(cfgfile) for cfgfile in successfully_read_files]}")
```
The reason this was changed was to make the user see only from where the configuration file was read and not the list of all configuration files.

Aathish04 replied:
`successfully_read_files` will only contain the cfg files that were, well, successfully read. Not all of the config files that are there.

What's the point of only showing the end user the last config file that was read?

They should know what files are being read and what changes each might make to the final config.

leotrs replied:
Agreed with @Aathish04

leotrs replied:
BTW, I know I reviewed and accepted this change in a previous PR, but now that I think of it, I think it should be reverted, as Aathish is doing now.

naveen521kk commented on manim/utils/config_utils.py:
```diff
@@ -18,6 +18,7 @@
 
 __all__ = ["_run_config", "_paths_config_file", "_from_command_line"]
```
NitPick: You missed `curr_config_dict` here.

Aathish04 replied:
Good catch! I'll add it in.

Aathish04 commented:
@naveen521kk 

> 1. Issue: This should not happen we need a error message here.
>    ![err](https://camo.githubusercontent.com/d97e54eb18d2cb074dcc4be34681fecce522c859/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3631323730313732393231323835383430382f3733363130343835323339363034383339342f706f772e504e47)

Instead, the program now asks the user if they want to export to the same directory they are reading from.


> 2. Suggestion: Move the ones, like the CLI in Green to centre

Eh, it doesn't look very nice. I'll let you be the judge:

![Untitled 2](https://user-images.githubusercontent.com/33193764/88369329-64a82980-cdad-11ea-8849-bd6642904bab.png)

I haven't committed that change.

> 3. Issue: Change the colour of the link to something else other then blue.

Yeah, shame powershell has a blue background. I have changed the link to green in the latest commit.

> 4. Suggestion: Read Configuration Files should not be shown while writing cfg files
This is a byproduct of the fact that the config files are read before anything else happens when the command `manim cfg` is executed. Also, I don't see why this will pose a problem, as the config utilities can know the current config preferences.

> 5. Issue: Export command doesn't work
The command `manim cfg export` does work. I had mistakenly left out the `cfg` in the main comment for this PR. This has since been rectified.

> 6. Issue: Write Only what is edited into manim.cfg instead of writing everything.
I have not touched the configwriter function in this PR. It, by _default_ , writes the entire contents of `default.cfg` , and just changes the values as required.

> 7. Nitpick:Create the directory before writing the file.
I have added a warning and made sure the directory is created before writing the file, in the latest commit.

leotrs commented:
> Suggestion: Move the ones, like the CLI in Green to centre

Agree with @Aathish04, centered titles don't look great.

> Suggestion: Read Configuration Files should not be shown while writing cfg files

I think they should be shown.

> Issue: Write Only what is edited into manim.cfg instead of writing everything.

This is an interesting suggestion, and I'm on the fence about it. If you feel strongly about it @naveen521kk please open a new issue and let's keep it separate from this PR.

leotrs commented on manim/__main__.py:
```diff
@@ -150,23 +151,38 @@ def get_module(file_name):
-            if sound_on:
-                play_error_sound()
+    if hasattr(args,"subcommands"):
+        if "cfg" in args.subcommands:
```
Question: is it possible for `hasattr(args,"subcommands")` to be `True` but `"cfg" in args.subcommands` to be False? If so, we need an `else` clause here. If not, then this `if` statement seems unnecessary?

Aathish04 replied:
I was trying to future-proof this bit of code here. _Right now_, `cfg` is the only possible subcommand, and if `hasattr(args,"subcommands")` is true, (and it will only be true if argparse finds a subcommand), the subcommand _has_ to be `cfg`.


In the future, if/when more subcommands are added, we could just use `elifs` here and extend functionality as required.

leotrs commented on manim/__main__.py:
```diff
@@ -150,23 +151,38 @@ def get_module(file_name):
+                elif subcommand == "export":
+                    cfg_file_utils.export(args.dir)
+            else:
+                logger.error("NO ARGUMENT PROVIDED, exiting...")
```
Nitpick: no need to yell ;)

Aathish04 replied:
I'll change that lol. Just thought it would get the user's attention.

leotrs commented on manim/utils/cfg_file_utils.py:
```diff
@@ -1,28 +1,30 @@
 """
-cfgwriter.py
+cfg_file_utils.py
```
Comment: I agree this file should be changed, but I'd rather something more descriptive, that says "this is the file that is ran when the cfg subcommand is executed". It is also hard to distinguish between this and `config.py`. Something like `cfg_subcommand.py` for this file, though that's too long... [I'm rambling here...]

leotrs replied:
It also clashes with `config_utils.py`...

Aathish04 replied:
I've renamed it to `cfg_subcmds.py`. It's basically what you suggested, but who needs vowels anyway :P

leotrs commented on manim/utils/cfg_file_utils.py:
```diff
@@ -67,9 +69,8 @@ def replace_keys(default):
 
 
-def main():
+def write(level=None):
```
Suggestion: I think it'd be great to add a few more config options that have nothing to do with logging colors to this function in this PR.

Aathish04 replied:
Is it really necessary in _this_ PR ? 
Adding more stuff that can be modified by the `write` command won't be pretty hard in and of itself, but we'd need to do input validation on each and every thing the user will try and input. 

The `logging` section's input validation was pretty straightforward, @naveen521kk just fed it through rich's own `Style.parse` method and it checked whether the style was valid.

It'll be a bit harder to perform input validation on the other stuff, especially since the configparser stores them as strings and we can't even check to see if the datatypes will match.

While it _is_ possible, I personally think that it'll deviate from the purpose of this PR, which is to get a subcommand system up and running.
Of course, I'll go ahead with doing all the input validation stuff if you insist, but it'll take more time than I'd like before this PR can be merged.

leotrs replied:
The point is that if a user were to run the current `cfg write` command, they might the impression that it is only intended to change/write/modify logging-related configurations, and not all of the rest.

I do see your point about input validation and I do acknowledge that might get tricky.

How about, for this PR we add one more config option that has nothing to do with logging/rich/colors, and that is a Boolean? That way the input validation is fairly straightforward and we'll leave a blueprint for how to add more in the future. And the `cfg write` subcommand won't have only rich/color configs.

Aathish04 replied:
Alright, so I've added the ability to edit just those configs whose values are booleans in the latest commit. There's still a lot of room for improvement in the input validation for it, and code reusability and all that, but I think another Issue/PR pair should be the place to discuss that.

For now though, you can set stuff like `write_to_movie` and whatnot.

leotrs commented on manim/utils/cfg_file_utils.py:
```diff
@@ -1,28 +1,30 @@
 INTRO_INSTRUCTIONS = """[red]The default colour is used by the input statement.
 If left empty, the default colour will be used.[/red]
-[magenta] For a full list of styles, visit[/magenta] https://rich.readthedocs.io/en/latest/style.html"""
+[magenta] For a full list of styles, visit[/magenta] [green]https://rich.readthedocs.io/en/latest/style.html[/green]"""
```
Suggestion: an alternative is to change the background color instead of the face color. Not sure if that'd look good tho.

Aathish04 replied:
I think it might be a bit jarring, but that's just me.

leotrs commented on manim/utils/cfg_file_utils.py:
```diff
@@ -86,31 +87,71 @@ def main():
+            to_path = os.path.join(path,'manim.cfg')
+        console.print(f"Exported final Config at {from_path} to {to_path}.")
+    else:
+        console.print("Could NOT write config.", style="red bold")
```
Question: should this be an error instead of a console.print?

Aathish04 replied:
I don't think this qualifies as an "error", per se, since it only crops up in the event that the end user explicitly states that they don't want to perform the action.

_Also I didn't want to import logger just for this, since it's only going to be used in one place :)_

leotrs commented on manim/utils/config_utils.py:
```diff
@@ -29,9 +30,11 @@ def _parse_file_writer_config(config_parser, args):
-    fw_config["input_file"] = args.file
-    fw_config["scene_names"] = args.scene_names if args.scene_names is not None else []
-    fw_config["output_file"] = args.output_file
+    # Don't set these if the a subcommand is invoked
```
Issue: this comment is just repeating what the next line of code is saying. I suggest changing to "If a subcommand is given, manim will not render a video and thus input/output files are not needed".

leotrs replied:
Tip: don't say what the code is doing, say why it is doing it.

Aathish04 replied:
Thanks for the suggestion! I'll be more descriptive when writing comments from now on :)

leotrs commented on manim/utils/config_utils.py:
```diff
@@ -130,23 +133,46 @@ def _parse_cli(arg_list, input=True):
         epilog="Made with <3 by the manim community devs",
     )
+    min_argvs = 4 if "py" in sys.argv[0] else 2
+    if input and (len(sys.argv) > min_argvs-1 # If "manim" is not the only command
+        and any(a == item for a in sys.argv for item in NON_ANIM_UTILS) #non-anim exists
+        or len(sys.argv) == min_argvs-1):
```
Suggestion: Put all of this into a separate function for readability. Kind of like how it was done for `_from_command_line()`.

leotrs replied:
Comment: yes, this is hacky, but as long as it works I think it's fine. Perhaps open an issue after merging this flagging that this should/cloud be improved.

Aathish04 replied:
> Suggestion: Put all of this into a separate function for readability.

I cannot thank you enough for pushing me to do this. I've managed to get rid of an extra `if` statement, and the readability is improved by like 10 times!

leotrs commented on manim/utils/config_utils.py:
```diff
@@ -130,23 +133,46 @@ def _parse_cli(arg_list, input=True):
+        cfg_related = subparsers.add_parser('cfg')
+        cfg_subparsers = cfg_related.add_subparsers(dest="cfg_subcommand")
+
+        cfg_write = cfg_subparsers.add_parser('write')
+        cfg_write.add_argument(
+            "--level",
+            choices=["user", "cwd"],
+            default=None,
+            help="Specify if this config is for user or just the working directory."
+            )
+        cfg_show = cfg_subparsers.add_parser('show')
+
+        cfg_export = cfg_subparsers.add_parser("export")
+        cfg_export.add_argument("--dir",default=os.getcwd())
```
Comment: the variable names are confusing me... these are ArgParser objects, correct? I'd suggest something like `cfg_show_parser`.

leotrs replied:
Actually, do we even need to use `cfg_show` anywhere else in the code? If not, don't store it in a variable.

Aathish04 replied:
>  I'd suggest something like cfg_show_parser.

I've changed the names as suggested, and also stopped storing `cfg_show` as a variable in the latest commit. Thanks for pointing them out!

leotrs commented on manim/utils/config_utils.py:
```diff
@@ -130,23 +133,46 @@ def _parse_cli(arg_list, input=True):
-            default="",
-        )
+        # If the only command is manim, or if there are only rendering related commands
+        if len(sys.argv) < 2 or not(any(sys.argv[1] == item for item in NON_ANIM_UTILS)):
```
Suggestion: As before, perhaps put this condition in a separate function for readability.

Aathish04 replied:
In the newest commit, this isn't even necessary anymore, the function for the previous conditional takes care of it and is just `not`ed over.

leotrs commented on manim/utils/config_utils.py:
```diff
@@ -397,3 +433,7 @@ def _run_config():
     file_writer_config = _parse_file_writer_config(config_parser, args)
     return args, config_parser, file_writer_config, successfully_read_files
+
+def curr_config_dict():
+    config=_run_config()[1]
+    return {section: dict(config[section]) for section in config.sections()}
```
Question: this is not _really_ the actual config dictionary, is it? This is just the result of reading the cascading config files. I would call this the finalized config parser. The actual finalized global config dictionary can only be obtained after _parse_config has been called from `config.py`.

Aathish04 replied:
Good point. I've renamed it to `finalized_configs_dict()` in the latest commit, since it doesn't return an ArgumentParser, but it's contents in dictionary form.

Aathish04 commented:
@leotrs I think I've addressed everything you suggested. Please let me know if there's anything else!

leotrs commented:
Everything looks good to me now. Before merging, is there a way we can add some tests to this? I understand that this needs user input to work...

Aathish04 commented:
> Everything looks good to me now. Before merging, is there a way we can add some tests to this? I understand that this needs user input to work...

Writing a test for `show` and `export` should be straightforward.

Writing a test for `write` could possibly be done if we use the "reading from a file" trick that I used for `-` as filename, but I haven't tried that for an application like this, where each EOF results in a new `input()` statement.

Worst case scenario we'd need a text file to read from for every single config value...

XorUnison commented:
I've tried to roughly catch up with this development. But aside from Write/Show/Export can we also get Edit which opens the according file similar to what the -p flag does with render output?

Aathish04 commented:
> I've tried to roughly catch up with this development. But aside from Write/Show/Export can we also get Edit which opens the according file similar to what the -p flag does with render output?

Do you think it would go better as it's own command or as a flag for `write`? I personally think `manim cfg write --open` might look nicer.

Aathish04 commented:
@leotrs I'd like to direct your attention towards commits 1f3091c and 59d846c.

I have added a condition so that run_config now gets sys.argv passed to it not only if manim is called from the command line, but also if `run_config` is called my `manim/__main__`, which happens when a subcommand calls it. 
Without this, `python3 -m manim cfg export` will NOT read the current directory's `cfg`, but `manim cfg export` will, for some odd reason.

Do you think this will cause problems along the line?

XorUnison commented:
> Do you think it would go better as its own command or as a flag for `write`? I personally think `manim cfg write --open` might look nicer.

Not entirely sure... having it as a flag of `write` might be nice for the association of file creation, but the same thing might be the issue too. I would assume that `write` isn't just used to initialize, but also possibly to reset, as in overwrite the cfg, right? Using it like this to revert to default might be nice. But if so, it would be very bad to have `open` as a flag of the command that can reset the cfg instantly.

Aathish04 commented:
@XorUnison 

>  I would assume that `write` isn't just used to initialize, but also possibly to reset, as in overwrite the cfg, right?
Before `cfg_subcmds.py:write` is executed, all the config files that are viable are read anyway, and when `write()` is called, the _final config dict_'s values are what are carried over and set as "defaults".

For example, if you already have a `manim.cfg` in the directory, it's contents will be viewed as the default by `write`.

> Using it like this to revert to default might be nice.
I don't think there really is any reason to make `write` have a reset functionality, since one can just trash the `manim.cfg` file, where ever it is, and `default.cfg` would be the only successfully read cfg file, unless there's a user level cfg file.

> But if so, it would be very bad to have `open` as a flag of the command that can reset the cfg instantly.

Nothing is erased the moment you call `cfg write`. It's only after you go through all the processes of filling in your desired changes that the entire file is replaced with one containing your changes.

All things considered, we could just add a conditional that checks if `--open` is a flag passed, and completely sidestep everything above and just open the file in the users text editor...

XorUnison commented:
So let me get this straight then... you want to have the cfg editing basically also available in the terminal? If I got that right then a `--open` flag seems a lot more fitting of course.

Aathish04 commented:
> So let me get this straight then... you want to have the cfg editing basically also available in the terminal?

Yeah, this is how `write` is intended to work. You can type in the changes you want, and leave it blank if you want it to take the default value.

> If I got that right then a `--open` flag seems a lot more fitting of course.

Cool, so just to confirm, the `--open` flag should open, (and create if it doesn't exist), but not _modify_ the `manim.cfg` file in the default text editor, yes?

XorUnison commented:
> Cool, so just to confirm, the `--open` flag should open, (and create if it doesn't exist), but not _modify_ the `manim.cfg` file in the default text editor, yes?

I think so, yes. If users open it like that and make changes in the text editor that should be up to them. So after opening the write command can and should end itself without further action.

huguesdevimeux commented on manim/utils/cfg_subcmds.py:
```diff
@@ -0,0 +1,175 @@
+
+    config_paths = _paths_config_file() + [os.path.abspath("manim.cfg")]
+    if save_to_userpath.lower() == "y" or level=="user":
+        if not os.path.exists(os.path.abspath(os.path.join(config_paths[1], ".."))):
```
Nitpick (like, a real nitpick) : Consider using `os.pardir` instead of '`..`'

huguesdevimeux commented on tests/test_cli/test_cli.py:
```diff
@@ -15,6 +15,7 @@ def capture(command,instream=None):
 
 
 def test_help(python_version):
+    os.chdir(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))
```
**Nitpick** : Consider using something else if you want to go up three directories. Use `os.pardir`

Aathish04 replied:
@huguesdevimeux Done :) I have made it change back to the original directory in `test_cfg_subcmd` itself, so there is no reason to do it here,

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,44 @@
+
+def test_cfg_help(python_version):
+    """Test if Manim successfully adds configparsers when a subcommand is invoked."""
+    os.chdir(os.path.dirname(__file__))
```
**Issue** : as we have discussed on Discord, I don't think doing this is a good idea. 
It works right now as you set back the right `test_directory` in `test_cli.py` so tests will work fine as `test_cli `is ran directly after `test_cfg_subcmd` by pytest, but won't when we will add other tests that will run between `test_cli `and `test_cfg_subcmd`.
Why could you not execute a command like that in the test to move to the right directory : 
`cd <os.path.dirname(__file__)> && python -m manim cdg --help` ?
Shouldn't this work?

Aathish04 replied:
> It works right now as you set back the right test_directory in test_cli.py so tests will work fine as test_cli is ran directly after test_cfg_subcmd by pytest, but won't when we will add other tests that will run between test_cli and test_cfg_subcmd.

> Why could you not execute a command like that in the test to move to the right directory :
`cd <os.path.dirname(__file__)> && python -m manim cfg --help` ?

Well, yes, that would definitely work. My only gripe with that is that the same command will have to be added for all of the `cfg` subcommand tests separately, but that is not much of an issue.

huguesdevimeux commented on manim/utils/cfg_subcmds.py:
```diff
@@ -0,0 +1,187 @@
+        console.print(
+            """You are reading the config from the same directory you are exporting to.
+This means that the exported config will overwrite the config for this directory.
+Are you sure you want to continue?[y/n]""",
```
**Issue :** for a strange reason the `y/n` did not appear on my command prompt when it was between two `[` (see image)
![image](https://user-images.githubusercontent.com/36239975/88716647-a9222500-d11f-11ea-909a-ebeb859bdf39.png)

I fixed it by changing `[y/n]` by `(y/n)` : 
![image](https://user-images.githubusercontent.com/36239975/88716734-ca831100-d11f-11ea-8b75-dc9a41a2e8d5.png)
```suggestion
Are you sure you want to continue? (y/n)""",
```

Aathish04 replied:
Ohh I think this is because Rich considers anything within square brackets to be a style, and doesn't treat it as _text_.

Pinging @naveen521kk to see if this is true.

huguesdevimeux commented on manim/utils/cfg_subcmds.py:
```diff
@@ -0,0 +1,187 @@
+            to_path = os.path.join(path,'manim.cfg')
+        console.print(f"Exported final Config at {from_path} to {to_path}.")
+    else:
+        console.print("Could NOT write config.", style="red bold")
```
**suggestion (if-minor)** : Maybe be more explicit here. This could be misleading, as one could understand that it was a bug instead of the user that typed `n`.

Aathish04 replied:
Agreed. I've made it say `Aborted...` now.

huguesdevimeux commented on manim/utils/file_ops.py:
```diff
@@ -56,4 +58,19 @@ def get_sorted_integer_files(directory,
+            commands.append("xdg-open")
+        elif current_os.startswith("CYGWIN"):
+            commands.append("cygstart")
+        else:  # Assume macOS
```
**Nitpick (if-minor)** : For macOs (but you surely do know that already), name of the os is `Darwin`.
I think it would be better to change it into and if, and to raise an error or to do something by default if the os is not found (in theory it should never append, as platform.system() return `''` if the system is not found and only returns four values.

Aathish04 replied:
Thank you for the suggestion! I've made it raise an `OSError` if the OS is not recognisable.

huguesdevimeux commented on manim/utils/file_ops.py:
```diff
@@ -56,4 +58,19 @@ def get_sorted_integer_files(directory,
+        commands = []
+        if current_os == "Linux":
+            commands.append("xdg-open")
+        elif current_os.startswith("CYGWIN"):
```
**Question :** Can `platform.system()` returns such a thing ? I thought it only returns four things (Java,Darwin, Windows, Linux). See https://docs.python.org/3/library/platform.html#platform.system

Will this work ?

Aathish04 replied:
Yup. If the end user is using Python via a Cygwin installation, it will be something like `CYGWIN_NT=6.3`

https://stackoverflow.com/questions/32612235/what-is-os-name-on-cygwin

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,52 @@
+    """Test if Manim successfully adds configparsers when a subcommand is invoked."""
+    command = f"cd {this_folder} && {python_version} -m manim cfg --help"
+    out, err, exitcode = capture(command, use_shell=True)
+    assert exitcode == 0, f"The cfg subcommand is not working as intended."
```
**Issue :** Add err somewhere in the message. It will be way way more faster to debug, believe me.
```suggestion
    assert exitcode == 0, f"The cfg subcommand help is not working as intended.\nError : {err}"
```

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,52 @@
+    """Test if the `manim cfg export` command works as intended."""
+    command = f"cd {this_folder} && {python_version} -m manim cfg export --dir temp"
+    out, err, exitcode = capture(command, use_shell=True)
+    assert exitcode == 0
```
Same comment as above :
```suggestion
    assert exitcode == 0, f"The cfg subcommand export is not working as intended.\nError : {err}"
```

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,52 @@
+        instream=open(os.path.join(this_folder, "write_cfg_sbcmd_input.txt")),
+        use_shell=True
+        )
+    assert exitcode == 0, err
```
```suggestion
    assert exitcode == 0, f"The cfg subcommand write is not working as intended.\nError : {err}"
`
```

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,52 @@
+    with open(cfgfilepath,"r") as cfgfile:
+        assert "sound = False" in cfgfile.read()
+
+    with open(cfgfilepath,"w") as cfgfile:
```
**Issue (blocking):** Ok, here, this is quite nitpicking. This can cause trouble when the test fail and nothing (or something wrong is written within the config file. As the test would have failed, this line will never be run and we will end up having a wrong config file for tests. 
You can fix that by using a pytest fixture : Here are two links that can help you : https://stackoverflow.com/questions/56997702/how-can-i-run-cleanup-code-only-for-specific-tests-with-pytest
https://docs.pytest.org/en/latest/fixture.html#using-fixtures-from-classes-modules-or-projects

(Don't hesitate to ask if we need help (although I'm not an expert on that))

Aathish04 replied:
Thank you for providing those resources. I've written a `reset_cfg_file` fixture that is used in the test now.

huguesdevimeux commented on tests/test_cli/test_cfg_subcmd.py:
```diff
@@ -0,0 +1,52 @@
+
+def test_cfg_help(python_version):
+    """Test if Manim successfully adds configparsers when a subcommand is invoked."""
+    command = f"cd {this_folder} && {python_version} -m manim cfg --help"
```
**Praise** : This is really a nice workaround. this is way better than changing the CWD :D

Aathish04 replied:
Thank you! I wouldn't have tried that if you didn't suggest it, so the credit is all yours :P

Aathish04 commented:
@leotrs @naveen521kk  @huguesdevimeux @XorUnison 

I think I've addressed all your recommendations. Feel free to tell me if there's anything else!

leotrs commented:
On my first review, I asked how can we add tests for this. Were those tests added? If not, please open a separate issue to that effect. Merging now!

huguesdevimeux commented:
> On my first review, I asked how can we add tests for this. Were those tests added? If not, please open a separate issue to that effect. ~Merging now!~
> 
> EDIT: tried to merge but there are conflicts. @Aathish04 please resolve conflicts and feel free to merge.

no need to create an issue. Add the test required in #193

Aathish04 commented:
> On my first review, I asked how can we add tests for this. Were those tests added? If not, please open a separate issue to that effect. ~Merging now!~
> 
> EDIT: tried to merge but there are conflicts. @Aathish04 please resolve conflicts and feel free to merge.

Yes, the tests mentioned were added.
